File: functions/src/services/weekResetService.ts
import { Firestore } from 'firebase-admin/firestore';
import { Task, TaskState, effectiveAssignedTo } from '../models/task';
import { Person } from '../models/person';
import { Flat } from '../models/flat';
import { TaskRepository } from '../repository/taskRepository';
import { PersonRepository } from '../repository/personRepository';
import { FlatRepository } from '../repository/flatRepository';
import {
  buildWeekResetContext,
  BlueShortVacationStrategy,
  GreenL3Strategy,
  GreenL2Strategy,
  RedL3Strategy,
  RedL2Strategy,
  RedL1Strategy,
  GreenL1Strategy,
  BlueLongVacationStrategy,
  AssignmentStrategy,
  WeekResetContext,
} from './assignmentStrategy';
import { LOG_WEEK_RESET_COMPLETE, LOG_WEEK_RESET_START } from '../constants/strings';
import * as logger from 'firebase-functions/logger';

/**
 * Orchestrates the weekly task reassignment algorithm (Template Method pattern).
 *
 * week_reset() must run as an atomic Firestore transaction to prevent partial state
 * (e.g. two people assigned to the same task) under concurrent execution or crashes.
 *
 * Step order matches the spec in CLAUDE.md:
 *   1. Blue short vacation (protected slots)
 *   2. Green L3 → L2 (scan forward)
 *   3. Green L2 → L1 (scan forward)
 *   4. Red L3 → stay L3
 *   5. Red L2 → up to L3
 *   6. Red L1 → up to L2
 *   7. Green L1 → fill remaining (shortest ring distance)
 *   8. Blue long vacation → fill remaining (unprotected)
 */
export class WeekResetService {
  private readonly taskRepo: TaskRepository;
  private readonly personRepo: PersonRepository;
  private readonly flatRepo: FlatRepository;

  /** Ordered list of assignment strategies executed each reset cycle. */
  private readonly strategies: AssignmentStrategy[] = [
    new BlueShortVacationStrategy(),
    new GreenL3Strategy(),
    new GreenL2Strategy(),
    new RedL3Strategy(),
    new RedL2Strategy(),
    new RedL1Strategy(),
    new GreenL1Strategy(),
    new BlueLongVacationStrategy(),
  ];

  constructor(db: Firestore) {
    this.taskRepo = new TaskRepository(db);
    this.personRepo = new PersonRepository(db);
    this.flatRepo = new FlatRepository(db);
  }

  /**
   * Runs the full weekly reset for a flat inside a single Firestore transaction.
   *
   * @param flatId - Firestore document ID of the flat.
   */
  async weekReset(flatId: string): Promise<void> {
    logger.info(LOG_WEEK_RESET_START, { flatId });

    const db = (this.taskRepo as unknown as { db: Firestore }).db;

    await db.runTransaction(async (transaction) => {
      // ── Load all data inside the transaction ─────────────────────────────
      const flat = await this.flatRepo.getFlatInTransaction(flatId, transaction);
      const tasks = await this.taskRepo.getAllTasksInTransaction(flatId, transaction);
      const persons = await this.personRepo.getAllMembersInTransaction(flatId, transaction);

      // ── Pre-step: increment weeks_not_cleaned ────────────────────────────
      const updatedTasks = this.incrementWeeksNotCleaned(tasks, persons);

      // ── Build context and run strategies ────────────────────────────────
      const ctx = buildWeekResetContext(
        updatedTasks,
        persons,
        flat.vacation_threshold_weeks,
      );
      for (const strategy of this.strategies) {
        strategy.execute(ctx);
      }

      // ── Write results back ───────────────────────────────────────────────
      this.writeResetResults(flatId, updatedTasks, persons, ctx, transaction);
    });

    logger.info(LOG_WEEK_RESET_COMPLETE, { flatId });
  }

  /**
   * Increments weeks_not_cleaned on every task whose effective assignee is on
   * vacation or whose state is Vacant. Returns a new array with updated tasks.
   * This happens before categorisation so the threshold comparison is current.
   */
  private incrementWeeksNotCleaned(tasks: Task[], persons: Person[]): Task[] {
    const vacationUids = new Set(
      persons.filter((p) => p.on_vacation).map((p) => p.uid),
    );

    return tasks.map((task) => {
      const effectiveUid = effectiveAssignedTo(task);
      const isVacant = task.state === TaskState.Vacant;
      const assigneeOnVacation = effectiveUid !== '' && vacationUids.has(effectiveUid);

      if (isVacant || assigneeOnVacation) {
        return { ...task, weeks_not_cleaned: task.weeks_not_cleaned + 1 };
      }
      return task;
    });
  }

  /**
   * Applies all computed next-week assignments and resets task/person state
   * in the transaction batch.
   *
   * Post-step actions (per spec):
   * - Assign each task to the person in ctx.nextAssignments
   * - Clear original_assigned_to on all tasks
   * - Set all task states to Pending
   * - Clear on_vacation on all persons
   */
  private writeResetResults(
    flatId: string,
    tasks: Task[],
    persons: Person[],
    ctx: WeekResetContext,
    transaction: FirebaseFirestore.Transaction,
  ): void {
    // Build ring_index → task.id map for write operations
    const taskByRingIndex = new Map<number, Task>();
    for (const task of tasks) {
      taskByRingIndex.set(task.ring_index, task);
    }

    // Update each task with its new assignment and reset state
    for (let ringIndex = 0; ringIndex < ctx.nextAssignments.length; ringIndex++) {
      const newAssignee = ctx.nextAssignments[ringIndex];
      const task = taskByRingIndex.get(ringIndex);
      if (!task) continue;

      this.taskRepo.updateTaskInTransaction(
        flatId,
        task.id,
        {
          assigned_to: newAssignee,
          original_assigned_to: '',
          state: TaskState.Pending,
          weeks_not_cleaned: task.weeks_not_cleaned, // already updated in incrementWeeksNotCleaned
        },
        transaction,
      );
    }

    // Clear on_vacation on all persons
    for (const person of persons) {
      this.personRepo.updateMemberInTransaction(
        flatId,
        person.uid,
        { on_vacation: false },
        transaction,
      );
    }
  }

  /**
   * Marks a task as completed. Resets weeks_not_cleaned and clears on_vacation
   * on the assigned person.
   * Does NOT modify original_assigned_to (swap tracking is independent).
   *
   * @param flatId - The flat's Firestore document ID.
   * @param taskId - The task's Firestore document ID.
   */
  async completedTask(flatId: string, taskId: string): Promise<void> {
    const db = (this.taskRepo as unknown as { db: Firestore }).db;

    await db.runTransaction(async (transaction) => {
      const task = await this.taskRepo.getTaskInTransaction(flatId, taskId, transaction);

      this.taskRepo.updateTaskInTransaction(
        flatId,
        taskId,
        { state: TaskState.Completed, weeks_not_cleaned: 0 },
        transaction,
      );

      // Clear vacation status on the assigned person (post-vacation return)
      const uid = effectiveAssignedTo(task);
      if (uid !== '') {
        this.personRepo.updateMemberInTransaction(
          flatId,
          uid,
          { on_vacation: false },
          transaction,
        );
      }
    });
  }

  /**
   * Records a swap request from `requesterUid` targeting `targetTaskId`.
   * Returns the created swap request document ID.
   */
  async requestChangeTask(
    flatId: string,
    requesterUid: string,
    requesterTaskId: string,
    targetTaskId: string,
    db: Firestore,
  ): Promise<string> {
    const { Timestamp } = await import('firebase-admin/firestore');
    const { COLLECTION_FLATS, COLLECTION_SWAP_REQUESTS } = await import('../constants/strings');
    const { SwapRequestStatus } = await import('../models/swapRequest');

    const swapRef = db
      .collection(COLLECTION_FLATS)
      .doc(flatId)
      .collection(COLLECTION_SWAP_REQUESTS)
      .doc();

    await swapRef.set({
      requester_uid: requesterUid,
      requester_task_id: requesterTaskId,
      target_task_id: targetTaskId,
      status: SwapRequestStatus.Pending,
      created_at: Timestamp.now(),
    });

    return swapRef.id;
  }

  /**
   * Accepts a swap request: swaps assigned_to on both tasks and deducts 1 token
   * from the requester.
   *
   * @throws When the request is not in Pending state or the requester has no tokens.
   */
  async acceptSwap(flatId: string, swapRequestId: string, db: Firestore): Promise<void> {
    const { COLLECTION_FLATS, COLLECTION_SWAP_REQUESTS, ERROR_SWAP_NOT_PENDING, ERROR_INSUFFICIENT_SWAP_TOKENS } =
      await import('../constants/strings');
    const { SwapRequestStatus, swapRequestFromFirestore } = await import('../models/swapRequest');

    await db.runTransaction(async (transaction) => {
      const swapRef = db
        .collection(COLLECTION_FLATS)
        .doc(flatId)
        .collection(COLLECTION_SWAP_REQUESTS)
        .doc(swapRequestId);

      const swapDoc = await transaction.get(swapRef);
      const swap = swapRequestFromFirestore(swapDoc.id, swapDoc.data()!);

      if (swap.status !== SwapRequestStatus.Pending) {
        throw new Error(ERROR_SWAP_NOT_PENDING);
      }

      const requester = await this.personRepo.getMember(flatId, swap.requester_uid);
      if (requester.swap_tokens_remaining <= 0) {
        throw new Error(ERROR_INSUFFICIENT_SWAP_TOKENS);
      }

      const requesterTask = await this.taskRepo.getTaskInTransaction(
        flatId,
        swap.requester_task_id,
        transaction,
      );
      const targetTask = await this.taskRepo.getTaskInTransaction(
        flatId,
        swap.target_task_id,
        transaction,
      );

      // Swap assigned_to; original_assigned_to tracks the pre-swap assignment
      this.taskRepo.updateTaskInTransaction(
        flatId,
        swap.requester_task_id,
        {
          assigned_to: targetTask.assigned_to,
          original_assigned_to: requesterTask.assigned_to,
        },
        transaction,
      );
      this.taskRepo.updateTaskInTransaction(
        flatId,
        swap.target_task_id,
        {
          assigned_to: requesterTask.assigned_to,
          original_assigned_to: targetTask.assigned_to,
        },
        transaction,
      );

      // Deduct one token from the requester
      this.personRepo.updateMemberInTransaction(
        flatId,
        swap.requester_uid,
        { swap_tokens_remaining: requester.swap_tokens_remaining - 1 },
        transaction,
      );

      // Mark request as accepted
      transaction.update(swapRef, { status: SwapRequestStatus.Accepted });
    });
  }
}

// ── Exported helper: builds context for testing without Firestore ─────────────

/**
 * Runs the week_reset algorithm on in-memory data without touching Firestore.
 * Used for unit and BDD tests.
 */
export function runWeekResetAlgorithm(
  tasks: Task[],
  persons: Person[],
  flat: Pick<Flat, 'vacation_threshold_weeks'>,
): string[] {
  // Simulate incrementWeeksNotCleaned
  const vacationUids = new Set(
    persons.filter((p) => p.on_vacation).map((p) => p.uid),
  );
  const updatedTasks = tasks.map((task) => {
    const effectiveUid = effectiveAssignedTo(task);
    const isVacant = task.state === TaskState.Vacant;
    const assigneeOnVacation = effectiveUid !== '' && vacationUids.has(effectiveUid);
    if (isVacant || assigneeOnVacation) {
      return { ...task, weeks_not_cleaned: task.weeks_not_cleaned + 1 };
    }
    return task;
  });

  const ctx: WeekResetContext = buildWeekResetContext(
    updatedTasks,
    persons,
    flat.vacation_threshold_weeks,
  );

  const strategies: AssignmentStrategy[] = [
    new BlueShortVacationStrategy(),
    new GreenL3Strategy(),
    new GreenL2Strategy(),
    new RedL3Strategy(),
    new RedL2Strategy(),
    new RedL1Strategy(),
    new GreenL1Strategy(),
    new BlueLongVacationStrategy(),
  ];

  for (const strategy of strategies) {
    strategy.execute(ctx);
  }

  return ctx.nextAssignments;
}

File: functions_python/services/week_reset_service.py
"""WeekResetService: orchestrates the weekly task reassignment algorithm.

Uses the Template Method pattern: a fixed sequence of AssignmentStrategy steps
runs inside a single Firestore transaction to guarantee atomicity.
"""

from __future__ import annotations

import logging
from typing import Any, ClassVar, cast

from constants.strings import (
    COLLECTION_FLATS,
    COLLECTION_SWAP_REQUESTS,
    ERROR_INSUFFICIENT_SWAP_TOKENS,
    ERROR_SWAP_NOT_PENDING,
    LOG_WEEK_RESET_COMPLETE,
    LOG_WEEK_RESET_START,
)
from models.person import Person
from models.swap_request import SwapRequestStatus, swap_request_from_firestore
from models.task import Task, TaskState, effective_assigned_to
from repository.flat_repository import FlatRepository
from repository.person_repository import PersonRepository
from repository.task_repository import TaskRepository
from services.assignment_strategy import (
    AssignmentStrategy,
    BlueLongVacationStrategy,
    BlueShortVacationStrategy,
    GreenL1Strategy,
    GreenL2Strategy,
    GreenL3Strategy,
    RedL1Strategy,
    RedL2Strategy,
    RedL3Strategy,
    WeekResetContext,
    build_week_reset_context,
)

logger = logging.getLogger(__name__)


class WeekResetService:
    """Orchestrates the weekly task reassignment algorithm.

    Step order matches the spec in CLAUDE.md:
      1. Blue short vacation (protected slots)
      2. Green L3 → L2/L1 (scan forward)
      3. Green L2 → L1 (scan forward)
      4. Red L3 → stay L3
      5. Red L2 → up to L3
      6. Red L1 → up to L2/L3
      7. Green L1 → fill remaining (shortest ring distance)
      8. Blue long vacation → fill remaining (unprotected)
    """

    # Fixed ordered list of strategies executed each reset cycle.
    _STRATEGIES: ClassVar[list[AssignmentStrategy]] = [
        BlueShortVacationStrategy(),
        GreenL3Strategy(),
        GreenL2Strategy(),
        RedL3Strategy(),
        RedL2Strategy(),
        RedL1Strategy(),
        GreenL1Strategy(),
        BlueLongVacationStrategy(),
    ]

    def __init__(self, db: Any) -> None:
        self._db = db
        self._task_repo = TaskRepository(db)
        self._person_repo = PersonRepository(db)
        self._flat_repo = FlatRepository(db)

    def week_reset(self, flat_id: str) -> None:
        """Run the full weekly reset for a flat inside a single Firestore transaction."""
        from google.cloud.firestore_v1.transaction import transactional

        logger.info("%s %s", LOG_WEEK_RESET_START, flat_id)

        @transactional
        def _run(transaction: Any) -> None:
            flat = self._flat_repo.get_flat_in_transaction(flat_id, transaction)
            tasks = self._task_repo.get_all_tasks_in_transaction(flat_id, transaction)
            persons = self._person_repo.get_all_members_in_transaction(flat_id, transaction)

            updated_tasks = _increment_weeks_not_cleaned(tasks, persons)

            ctx = build_week_reset_context(updated_tasks, persons, flat.vacation_threshold_weeks)
            for strategy in self._STRATEGIES:
                strategy.execute(ctx)

            _write_reset_results(
                flat_id,
                updated_tasks,
                ctx,
                transaction,
                self._task_repo,
            )

        _run(self._db.transaction())
        logger.info("%s %s", LOG_WEEK_RESET_COMPLETE, flat_id)

    def completed_task(self, flat_id: str, task_id: str) -> None:
        """Mark a task as completed and clear the assignee's on_vacation flag.

        Does NOT modify original_assigned_to — swap tracking is independent.
        """
        from google.cloud.firestore_v1.transaction import transactional

        @transactional
        def _run(transaction: Any) -> None:
            task = self._task_repo.get_task_in_transaction(flat_id, task_id, transaction)
            self._task_repo.update_task_in_transaction(
                flat_id,
                task_id,
                {"state": TaskState.Completed.value, "weeks_not_cleaned": 0},
                transaction,
            )
            uid = effective_assigned_to(task)
            if uid != "":
                self._person_repo.update_member_in_transaction(flat_id, uid, {"on_vacation": False}, transaction)

        _run(self._db.transaction())

    def request_change_task(
        self,
        flat_id: str,
        requester_uid: str,
        requester_task_id: str,
        target_task_id: str,
    ) -> str:
        """Record a swap request from requester_uid targeting target_task_id.

        Returns the created swap request document ID.
        """
        from google.cloud.firestore_v1 import SERVER_TIMESTAMP

        swap_ref = (
            self._db.collection(COLLECTION_FLATS).document(flat_id).collection(COLLECTION_SWAP_REQUESTS).document()
        )
        swap_ref.set(
            {
                "requester_uid": requester_uid,
                "requester_task_id": requester_task_id,
                "target_task_id": target_task_id,
                "status": SwapRequestStatus.Pending.value,
                "created_at": SERVER_TIMESTAMP,
            }
        )
        return str(swap_ref.id)

    def accept_swap(self, flat_id: str, swap_request_id: str) -> None:
        """Accept a swap request: swap assigned_to on both tasks, deduct 1 token.

        Raises ValueError when the request is not Pending or the requester has no tokens.
        """
        from google.cloud.firestore_v1.transaction import transactional

        swap_ref = (
            self._db.collection(COLLECTION_FLATS)
            .document(flat_id)
            .collection(COLLECTION_SWAP_REQUESTS)
            .document(swap_request_id)
        )

        @transactional
        def _run(transaction: Any) -> None:
            swap_doc = swap_ref.get(transaction=transaction)
            swap = swap_request_from_firestore(swap_doc.id, swap_doc.to_dict())

            if swap.status != SwapRequestStatus.Pending:
                raise ValueError(ERROR_SWAP_NOT_PENDING)

            requester = self._person_repo.get_member(flat_id, swap.requester_uid)
            if requester.swap_tokens_remaining <= 0:
                raise ValueError(ERROR_INSUFFICIENT_SWAP_TOKENS)

            requester_task = self._task_repo.get_task_in_transaction(flat_id, swap.requester_task_id, transaction)
            target_task = self._task_repo.get_task_in_transaction(flat_id, swap.target_task_id, transaction)

            self._task_repo.update_task_in_transaction(
                flat_id,
                swap.requester_task_id,
                {
                    "assigned_to": target_task.assigned_to,
                    "original_assigned_to": requester_task.assigned_to,
                },
                transaction,
            )
            self._task_repo.update_task_in_transaction(
                flat_id,
                swap.target_task_id,
                {
                    "assigned_to": requester_task.assigned_to,
                    "original_assigned_to": target_task.assigned_to,
                },
                transaction,
            )
            self._person_repo.update_member_in_transaction(
                flat_id,
                swap.requester_uid,
                {"swap_tokens_remaining": requester.swap_tokens_remaining - 1},
                transaction,
            )
            transaction.update(swap_ref, {"status": SwapRequestStatus.Accepted.value})

        _run(self._db.transaction())


# ── Private helpers ───────────────────────────────────────────────────────────


def _increment_weeks_not_cleaned(tasks: list[Task], persons: list[Person]) -> list[Task]:
    """Increment weeks_not_cleaned on every task whose assignee is on vacation or Vacant.

    Returns a new list with updated tasks; happens before categorisation so the
    threshold comparison uses the current value.
    """
    vacation_uids = {p.uid for p in persons if p.on_vacation}
    result: list[Task] = []
    for task in tasks:
        effective_uid = effective_assigned_to(task)
        is_vacant = task.state == TaskState.Vacant
        assignee_on_vacation = effective_uid != "" and effective_uid in vacation_uids
        if is_vacant or assignee_on_vacation:
            from dataclasses import replace

            result.append(replace(task, weeks_not_cleaned=task.weeks_not_cleaned + 1))
        else:
            result.append(task)
    return result


def _write_reset_results(
    flat_id: str,
    tasks: list[Task],
    ctx: WeekResetContext,
    transaction: Any,
    task_repo: TaskRepository,
) -> None:
    """Apply all computed next-week assignments and reset task/person state in the transaction."""
    task_by_ring_index = {task.ring_index: task for task in tasks}

    for ring_index, new_assignee in enumerate(ctx.next_assignments):
        task = task_by_ring_index.get(ring_index)
        if task is None:
            continue
        task_repo.update_task_in_transaction(
            flat_id,
            task.id,
            {
                "assigned_to": new_assignee,
                "original_assigned_to": "",
                "state": TaskState.Pending.value,
                "weeks_not_cleaned": task.weeks_not_cleaned,
            },
            transaction,
        )

    # on_vacation is intentionally NOT cleared here.
    # Per spec, only completed_task() clears on_vacation — when a person
    # completes their assigned task they are considered back from vacation.
    # Clearing it on reset would show vacation-assigned tasks as yellow
    # instead of blue.


# ── Exported helper: pure in-memory algorithm for testing ────────────────────


def run_week_reset_algorithm(
    tasks: list[Task],
    persons: list[Person],
    flat: Any,  # Any object with vacation_threshold_weeks attribute or dict key
) -> list[str]:
    """Run the week_reset algorithm on in-memory data without touching Firestore.

    Used for unit and BDD tests. Returns next_assignments as a list[str].
    """
    threshold = flat["vacation_threshold_weeks"] if isinstance(flat, dict) else flat.vacation_threshold_weeks

    updated_tasks = _increment_weeks_not_cleaned(tasks, persons)
    ctx = build_week_reset_context(updated_tasks, persons, threshold)

    strategies: list[AssignmentStrategy] = [
        BlueShortVacationStrategy(),
        GreenL3Strategy(),
        GreenL2Strategy(),
        RedL3Strategy(),
        RedL2Strategy(),
        RedL1Strategy(),
        GreenL1Strategy(),
        BlueLongVacationStrategy(),
    ]
    for strategy in strategies:
        strategy.execute(ctx)

    return cast(list[str], ctx.next_assignments)
